README.md
# CS:GO Market API implementation
**CS:GO Markte API** - this module is a Python client library for CS:GO Market (CSGO.TM before) platform API (CS:GO Markte API wrapper)

**CS:GO Market** - one of the largest stores of virtual items from the CS:GO

**Market** - https://market.csgo.com/

**API Documentation** - https://market.csgo.com/docs-v2

# Installation

Install current version with [PyPi]():

```bash
pip install csgotm-py==0.1.0
```


Or from GitHub:

```bash
pip install https://github.com/wendland0d/csgotm-py/archive/refs/heads/main.zip
```

ChangeLog.md
# v0.1.0
- Account API
- Setting up initial requirements
- "First commit"
- Uploaded on PyPi

# v0.2.0

- Refactored the structure of module
- Added exceptions
- API:
    1. Added trades methods 
    2. Changed some methods with new exceptions

csgotm/account.py
import requests

from .market import Market

class Account(Market):

    AVAILABE_CURRENCY = ['RUB', 'USD', 'EUR']
    INCORRECT_CURRENCY = f'Неправильно выбрана валюта. Доступные сейчас {AVAILABE_CURRENCY}'


    def __init__(self, token: str) -> None:
        super().__init__(token)


    # POST Method 'ping'
    def ping(self):
        response = requests.post(f'https://market.csgo.com/api/v2/ping?key={self.token}').json()

        if response['success'] == True:
            print('PONG! Продажи включены')
        else:
            print(self.ERROR_RESPONSE)        


    # POST Method 'get_money' gives back your currency and amount of money on your account
    def get_money(self):
        response = requests.post(f'https://market.csgo.com/api/v2/get-money?key={self.token}').json()

        if response['success'] == False:
            print('Данные недоступны и/или отсутствуют. Проверьте API-ключ')
        else:
            print(f'Валюта: {response["currency"]}\nБаланс: {response["money"]}')


    # POST Metgod 'go_offline' turns down your sales
    def go_offline(self):
        response = requests.post(f'https://market.csgo.com/api/v2/go-offline?key={self.token}').json()

        if response['success'] == True:
            print('Продажи остановлены!')
        else:
            print(self.ERROR_RESPONSE)


    # POST Method 'update_inventory' refreshes the inventory cache (recommended to do after each accepted trade offer)
    def update_inventory(self):
        response = requests.post(f'https://market.csgo.com/api/v2/update-inventory/?key={self.token}').json()

        if response['success'] == True:
            print('Инвентарь обновлен!')
        else:
            print(self.ERROR_RESPONSE)


    # REMEMBER! THIS METHOD IS USELESS STILL DISCOUNT PROGRAMM REWORKING!
    # POST Method 'transef_discount' moves your discount to another account. 
    def transef_discount(self, his_secret_key: str):
        response = requests.post(f'https://market.csgo.com/api/v2/transfer-discounts?key={self.token}&to={his_secret_key}').json()

        if response['success'] == True:
            print('Скидка перенесена!')
        else:
            print(self.ERROR_RESPONSE)


    # POST Method 'get_my_steam_id' gives your STEAMID 32/64 
    def get_my_steam_id(self):
        response = requests.post(f'https://market.csgo.com/api/v2/get-my-steam-id?key={self.token}').json()

        if response['success'] == True:
            return print(f'SteamID32: {response["steamid32"]}\nSteamID64: {response["steamid64"]}')
        else:
            return print(self.ERROR_RESPONSE)


    # POST Method 'set_pay_password' is used to set-up your new payment password. 
    # Warning - if you have never use payment password - go through 'https://market.csgo.com/checkout/password' to make it once.
    def set_pay_password(self, new_password: str = None):
        if new_password == None or self.payment_password == None:
            return print('Вы не установили платежный пароль и/или не ввели новый платежный пароль')
        
        response = requests.post(f'https://market.csgo.com/api/v2/set-pay-password?old_password={self.payment_password}&new_password={new_password}&key={self.token}').json()

        if response['success'] == True:
            self.payment_password = new_password
            print('Установлен новый платежный пароль!')
        else:
            print(self.ERROR_RESPONSE)


    # POST Method 'money_send' used to transfer balance from one to another account.
    # Don't forget to choose correct amount of money (1 RUB = 100, 1 USD = 1000, 1 EUR = 1000)
    def money_send(self, token_to: str, amount: str):
        if self.payment_password == None:
            return print(self.PAYMENT_PASSWORD_ERROR)
        
        response = requests.post(f'https://market.csgo.com/api/v2/money-send/{amount}/{token_to}?pay_pass={self.payment_password}&key={self.token}').json()
        curr = requests.post(f'https://market.csgo.com/api/v2/get-money?key={self.token}').json()['currency']

        if curr == 'RUB':
            amount_sep = 100
        else:
            amount_sep = 1000

        if response['success'] == True:
            return print(f'Успешно переведено {amount / amount_sep} {curr}!')
        else:
            return print(self.ERROR_RESPONSE)


    # POST Method 'money_send_history' shows transfer balance history. 
    def money_send_history(self, page: str = '0'):
        response = requests.post(f'https://market.csgo.com/api/v2/money-send-history/{page}?key={self.token}').json()

        checkup_list = [] # To Do - giga bicycle that needed to fix

        if response['success'] == True:
            if response['data'] == checkup_list:
                return print(self.NULL_HISTORY_ERROR)
            else:
                return print(f'Последний трансфер\n'
                             f'Куда - {response["data"][0]["to"]}\n'
                             f'Сколько - {response["data"][0]["amount_from"]} {response["data"][0]["currency_from"]}')
        else:
            return print(self.ERROR_RESPONSE)


    # POST Method 'set_steam_api_key' links your account w/ SteamAPI by Steam API Key
    def set_steam_api_key(self, steam_api_key: str):
        response = requests.post(f'https://market.csgo.com/api/v2/set-steam-api-key?key={self.token}&steam-api-key={steam_api_key}').json()

        if response['success'] == True:
            return print(f'Steam API Key {steam_api_key} успешно привязан к вашему аккаунту!')
        else:
            return print(self.INCORRECT_STEAM_KEY)


    # POST Method 'set_trade_link' sets up new trade link w/ your account.
    # You can find your link here -> http://steamcommunity.com/profiles/YOUR-STEAM-ID/tradeoffers/privacy.
    def set_trade_link(self, link: str):
        trade_token = link.split(sep='=')[2]

        response = requests.post(f'https://market.csgo.com/api/v2/set-trade-token?key={self.token}&token={trade_token}').json()

        if response['success'] == True:
            return print('Ссылка успешно установлена!')
        else:
            return print(self.INCORRECT_TRADE_LINK)


    # POST Method 'change_currency' change your account currency. By default in method 'new_currency' set 'USD'
    def change_currency(self, new_currency: str = 'USD'):
        response = requests.post(f'https://market.csgo.com/api/v2/change-currency/{new_currency}?key={self.token}')

        if response['success'] == True:
            return print(f'Валюта аккаунта успешно изменена!')
        else:
            if response['error'] == 1001: # ToDo: here and below should change to exceptions
                return print('Нельзя менять валюту, пока у Вас выставлены товары на продажу. Дождитесь когда их купят, или снимите с продажи.')
            if response['error'] == 1003:
                return print('Нельзя менять валюту, пока у Вас есть активные заявки на вывод. Дождитесь когда они будут исполнены, или отмените их вручную.')
            if response['error'] == 1004:
                return print('У Вас уже выбрана данная валюта')
            if response['error'] == 1005:
                return print('Выбрана неверная валюта')
            if response['error'] == 1007:
                return print('Ошибка. Вы два или более раз попытались отправить запрос на смену валюты.')    

csgotm/exceptions.py
class CurrencyError(Exception):
    def __init__(self, text: str = 'Wrong currency selected') -> None:
        self.text = text

class DataOrApiKeyError(Exception):
    def __init__(self, text: str = 'Data not available and/or missing. Check API Key') -> None:
        self.text = text

class EmptyPaymentPasswordError(Exception):
    def __init__(self, text: str = 'You have not set a payment password') -> None:
        self.text = text

class NewPaymentPasswordError(Exception):
    def __init__(self, text: str = 'You have not entered a new payment password') -> None:
        self.text = text

class NullHistoryError(Exception):
    def __init__(self, text: str = 'There are no transaction records in the history') -> None:
        self.text = text
        
class IncorectSteamKeyError(Exception):
    def __init__(self, text: str = 'Check if the Steam API key is correct') -> None:
        self.text = text

class InvalidSteamLinkError(Exception):
    def __init__(self, text: str = 'Check if the trade link is correct') -> None:
        self.text = text

class ItemsOnSaleError(Exception):
    def __init__(self, text: str = 'You cannot change the currency while you have goods for sale. Wait until they are bought, or withdraw from sale.') -> None:
        self.text = text
    
class ActiveWithdrawlsError(Exception):
    def __init__(self, text: str = 'You cannot change the currency while you have active withdrawal requests. Wait until they are executed, or cancel them manually.') -> None:
        self.text = text

class CurrencySelectedError(Exception):
    def __init__(self, text: str = 'You already have this currency selected') -> None:
        self.text = text

class CurrencyDoubleChangeError(Exception):
    def __init__(self, text: str = 'You tried to send a currency change request two or more times.') -> None:
        self.text = text

class EmptyBotTrade(Exception):
    def __init__(self, text: str = 'Nothing to send or receive') -> None:
        self.text = text

csgotm/market.py
from csgotm.exceptions import *

from typing import Union
import requests

class Market():

    """
    Main class for touch CS:GO Market API    
    """
    def __init__(self, token: str, payment_password: Union[str, None] = None) -> None:
        self.token = token
        self.payment_password = payment_password

    """
    Methods for getting in-site information about items, buy-orders and sell-orders
    """
    @staticmethod
    def show_prices(currency: str = "USD") -> dict:
        """
        Returns `dict` object contains all market items
        
        :param currency: Takes chosen currency into GET-request
        :raises csgotm.exceptions.market_exceptions.CurrencyError: If chosen currency that not available now
        """

        if currency not in ['RUB', 'USD', 'EUR']:
            raise CurrencyError()
        
        response = requests.get(f'https://market.csgo.com/api/v2/prices/{currency}.json').json()
        return response
        
    @staticmethod
    def show_class_instance_price(currency: str = "USD") -> dict:
        """
        Returns `dict` object contains all market items with extended inforrmtion per item

        :param currency: Takes chosen currency into GET-request
        :raises csgotm.exceptions.market_exceptions.CurrencyError: If chosen currency that not available now
        """

        if currency not in ['RUB', 'USD', 'EUR']:
            raise CurrencyError()
        
        response = requests.get(f'https://market.csgo.com/api/v2/prices/class_instance/{currency}.json').json()
        return response
    
    @staticmethod
    def show_buy_orders(currency: str = "USD") -> dict:
        if currency not in ['RUB', 'USD', 'EUR']:
            raise CurrencyError()
        
        response = requests.get(f'https://market.csgo.com/api/v2/prices/orders/{currency}.json').json()

        return response


    """
    Personal account methods 
    """
    def ping(self) -> dict:
        """
        Returns `dict` object contains ping-pong status of account

        :raises csgotm.exceptions.DataOrApiKeyError: If got invalid token
        """
        response = requests.post(f'https://market.csgo.com/api/v2/ping?key={self.token}').json()

        if response['success'] == True:
            return response
        else:
            raise DataOrApiKeyError() 

    def get_money(self) -> dict:
        response = requests.post(f'https://market.csgo.com/api/v2/get-money?key={self.token}').json()

        if response['success'] == True:
            return response
        else:
            raise DataOrApiKeyError()

    def go_offline(self) -> dict:
        response = requests.post(f'https://market.csgo.com/api/v2/go-offline?key={self.token}').json()

        if response['success'] == True:
            return response
        else:
            raise DataOrApiKeyError()
        
    def update_inventory(self) -> dict:
        """
        POST Method 'update_inventory' refreshes the inventory cache (recommended to do after each accepted trade offer)
        """

        response = requests.post(f'https://market.csgo.com/api/v2/update-inventory/?key={self.token}').json()

        if response['success'] == True:
            return response
        else:
            raise DataOrApiKeyError()

    def get_my_steam_id(self) -> dict:
        response = requests.post(f'https://market.csgo.com/api/v2/get-my-steam-id?key={self.token}').json()

        if response['success'] == True:
            return {'steamid32':response['steamid32'], 'steamid64':response['steamid64']}
        else:
            raise DataOrApiKeyError()
        
    def set_pay_password(self, new_password: Union[str, None] = None) -> dict:
        if self.payment_password == None:
            raise EmptyPaymentPasswordError()
        if new_password == None:
            raise NewPaymentPasswordError()
        
        response = requests.post(f'https://market.csgo.com/api/v2/set-pay-password?old_password={self.payment_password}&new_password={new_password}&key={self.token}').json()

        if response['success'] == True:
            self.payment_password = new_password
            return response
        else:
            return DataOrApiKeyError()
        
    def money_send(self, token_to: str, amount: str) -> str:
        if self.payment_password == None:
            raise EmptyPaymentPasswordError()
        
        response = requests.post(f'https://market.csgo.com/api/v2/money-send/{amount}/{token_to}?pay_pass={self.payment_password}&key={self.token}').json()
        curr = requests.post(f'https://market.csgo.com/api/v2/get-money?key={self.token}').json()['currency']

        if curr == 'RUB':
            amount_sep = 100
        else:
            amount_sep = 1000

        if response['success'] == True:
            return f'Successfully transferred {amount / amount_sep} {curr}!'
        else:
            raise DataOrApiKeyError()
        
    def money_send_history(self, page: str = '0', full_history: bool = False) -> dict:
            response = requests.post(f'https://market.csgo.com/api/v2/money-send-history/{page}?key={self.token}').json()

            checkup_list = [] # To Do - giga bicycle that needed to fix

            if response['success'] == True:
                if response['data'] == checkup_list:
                    raise NullHistoryError()
                else:
                    if full_history == False:
                        return {'To' : response['data'][0]['to'], 'Amount' : response['data'][0]['amount_from']}
                    else:
                        return response
            else:
                raise DataOrApiKeyError()
            
    def set_steam_api_key(self, steam_api_key: str) -> dict:
        response = requests.post(f'https://market.csgo.com/api/v2/set-steam-api-key?key={self.token}&steam-api-key={steam_api_key}').json()

        if response['success'] == True:
            return {'success': True, 'steam_api': steam_api_key}
        else:
            raise IncorectSteamKeyError()
        
    def set_trade_link(self, link: str) -> dict:
        trade_token = link.split(sep='=')[2]

        response = requests.post(f'https://market.csgo.com/api/v2/set-trade-token?key={self.token}&token={trade_token}').json()

        if response['success'] == True:
            return {'success': True, 'link': link}
        else:
            raise InvalidSteamLinkError()
        
    def change_currency(self, new_currency: str = 'USD') -> dict:
        if new_currency not in ['RUB', 'USD', 'EUR']:
            raise CurrencyError()

        response = requests.post(f'https://market.csgo.com/api/v2/change-currency/{new_currency}?key={self.token}')

        if response['success'] == True:
            return {'success': True, 'actual_currency': new_currency}
        else:
            if response['error'] == 1001: # ToDo: here and below should change to exceptions
                raise ItemsOnSaleError()
            if response['error'] == 1003:
                raise ActiveWithdrawlsError()
            if response['error'] == 1004:
                raise CurrencySelectedError()
            if response['error'] == 1005:
                raise CurrencyError()
            if response['error'] == 1007:
                raise CurrencyDoubleChangeError()
            
    """
    Account's methods. Trades methods
    """

    def trade_request_take(self, bot_id: Union[int, None] = None) -> dict:
        request_link = f'https://market.csgo.com/api/v2/trade-request-take?key={self.token}'
        if bot_id != None:
            request_link += f'&bot={bot_id}'

        response = requests.post(request_link).json()

        if response['success'] == True:
            return response
        else:
            raise EmptyBotTrade()
        
    def trade_request_give(self) -> dict:
        response = requests.post(f'https://market.csgo.com/api/v2/trade-request-give?key={self.token}').json

        if response['success'] == True:
            return response
        else:
            raise EmptyBotTrade()

    def trade_request_give_p2p(self) -> dict:
        response = requests.post(f'https://market.csgo.com/api/v2/trade-request-give-p2p?key={self.token}').json()

        return response
    
    def trade_request_give_p2p_all(self) -> dict:
        response = requests.post(f'https://market.csgo.com/api/v2/trade-request-give-p2p-all?key={self.token}')

        return response
    
